### ISSUE TITLE
Errors running ILSpy Console on Debian

### ISSUE THREAD

**Issue by chrisdoman** [icsharpcode/ILSpy#975]
Anyone else having issues running ILSpy Console on Debian?
I can get --help to print but it's not happy actually decompiling -

root@7997499eaa10:/shared/ILSpy-master/ICSharpCode.Decompiler.Console/bin/Release/netcoreapp2.0/debian-x64# ./ilspycmd ilspycmd.dll

Unhandled Exception: System.ArgumentOutOfRangeException: Index was out of range. Must be non-negative and less than the size of the collection.
Parameter name: index
   at System.ThrowHelper.ThrowArgumentOutOfRange_IndexException()
   at ICSharpCode.Decompiler.UniversalAssemblyResolver.GetAssemblyInNetGac(AssemblyNameReference reference, ReaderParameters parameters) in D:\GitWorkspace\ILSpy\ICSharpCode.Decompiler\DotNetCore\UniversalAssemblyResolver.cs:line 310
   at ICSharpCode.Decompiler.UniversalAssemblyResolver.ResolveInternal(AssemblyNameReference name, ReaderParameters parameters) in D:\GitWorkspace\ILSpy\ICSharpCode.Decompiler\DotNetCore\UniversalAssemblyResolver.cs:line 134
   at ICSharpCode.Decompiler.TypeSystem.DecompilerTypeSystem..ctor(ModuleDefinition moduleDefinition) in D:\GitWorkspace\ILSpy\ICSharpCode.Decompiler\TypeSystem\DecompilerTypeSystem.cs:line 57
   at ICSharpCode.Decompiler.Console.Program.Decompile(String assemblyFileName, TextWriter output, String typeName) in /shared/ILSpy-master/ICSharpCode.Decompiler.Console/Program.cs:line 96
   at ICSharpCode.Decompiler.Console.Program.<>c__DisplayClass0_0.<Main>b__0() in /shared/ILSpy-master/ICSharpCode.Decompiler.Console/Program.cs:line 70
   at McMaster.Extensions.CommandLineUtils.CommandLineApplication.Execute(String[] args)
   at ICSharpCode.Decompiler.Console.Program.Main(String[] args) in /shared/ILSpy-master/ICSharpCode.Decompiler.Console/Program.cs:line 73
Aborted

**Comment by christophwille** [comment]
This sounds similar to https://github.com/icsharpcode/ILSpy/issues/964#issuecomment-341368991 (Note that at that time cmdline & ps worked on Mac, but not in the context of Xamarin workbooks)

**Comment by siegfriedpammer** [comment]
Did you try `dotnet publish` or `dotnet run -- args`?

**Comment by tzachshabtay** [comment]
I'm getting the same exception on a Mac. This is when trying to decompile a mono application.
I took a brief look at the code: UniversalAssemblyResolver has an "OnMono" variable which checks whether or not the mono runtime type exists. So this tells whether the decompiler is compiled with mono or not, not whether the app decompiled is using Mono. 

So the answer would be no, the decompiler is compiled using dotnet core, not mono. 
The "GetGacPath", if we're not in Mono, assumes we're running on Windows and checks for the WINDIR environment variable, which does not exist on Mac (and Linux), and so assigns an empty list to gac_paths.

But then GetAssemblyInNetGac assumes that gac_paths has at least 2 members in it (?), and crashes.

So to sum up:
bug 1 -> `i < 2` in GetAssemblyInNetGac should actually by `i < gac_paths.length`.
bug 2 -> I think the OnMono variable should point to whether the app we decompile is mono or not, as we look for the assembly in the GAC, though I'm not sure if that variable also has other uses which justify it being the decompiler in mono?
bug 3 -> dotnet core GAC resolution is not there at all?

If somebody needs to repro, this is my project: https://github.com/tzachshabtay/MonoAGS, and I'm trying to decompile AGS.Engine.Desktop.dll.

**Comment by siegfriedpammer** [comment]
Thanks for your research... I had suspected something like that, but didn't "see" it until reading your post...

bug 1: Fixed
bug 2: Fixed - replaced with additional checks for `System.Environment.OSVersion.Platform == Unix`
bug 3: it's already there... see [UniversalAssemblyResolver.cs#L93](https://github.com/icsharpcode/ILSpy/blob/master/ICSharpCode.Decompiler/DotNetCore/UniversalAssemblyResolver.cs#L93) and [DotNetCorePathFinder.cs](https://github.com/icsharpcode/ILSpy/blob/master/ICSharpCode.Decompiler/DotNetCore/DotNetCorePathFinder.cs)

I hope this will solve this problem.

Thanks again!

**Comment by tzachshabtay** [comment]
@siegfriedpammer re bug 1, see my edit, you actually need 3 nested loops there, I believe.

**Comment by siegfriedpammer** [comment]
@chrisdoman @tzachshabtay I just pushed a fix for this... could you try the  build 3375 or higher?

**Comment by chrisdoman** [comment]
I'm afraid I'm still getting the same issue with the latest build, though It's possible Im just doing something silly here



root@161a8e6a2bd4:/shared/ilspy/ICSharpCode.Decompiler.Console# ls
ICSharpCode.Decompiler.Console.csproj  Program.cs  Running.gif
LICENSE				       README.md   TypesParser.cs
root@161a8e6a2bd4:/shared/ilspy/ICSharpCode.Decompiler.Console# dotnet publish -c Release -r debian-x64
Microsoft (R) Build Engine version 15.3.409.57025 for .NET Core
Copyright (C) Microsoft Corporation. All rights reserved.

  ICSharpCode.Decompiler.Console -> /shared/ilspy/ICSharpCode.Decompiler.Console/bin/Release/netcoreapp2.0/debian-x64/ilspycmd.dll
  ICSharpCode.Decompiler.Console -> /shared/ilspy/ICSharpCode.Decompiler.Console/bin/Release/netcoreapp2.0/debian-x64/publish/
root@161a8e6a2bd4:/shared/ilspy/ICSharpCode.Decompiler.Console# ls
ICSharpCode.Decompiler.Console.csproj  Program.cs  Running.gif	   bin
LICENSE				       README.md   TypesParser.cs  obj
root@161a8e6a2bd4:/shared/ilspy/ICSharpCode.Decompiler.Console# cd bin
root@161a8e6a2bd4:/shared/ilspy/ICSharpCode.Decompiler.Console/bin# ls
Release
root@161a8e6a2bd4:/shared/ilspy/ICSharpCode.Decompiler.Console/bin# cd Release/
root@161a8e6a2bd4:/shared/ilspy/ICSharpCode.Decompiler.Console/bin/Release# ls
netcoreapp2.0
root@161a8e6a2bd4:/shared/ilspy/ICSharpCode.Decompiler.Console/bin/Release# cd netcoreapp2.0/
root@161a8e6a2bd4:/shared/ilspy/ICSharpCode.Decompiler.Console/bin/Release/netcoreapp2.0# ls
debian-x64
root@161a8e6a2bd4:/shared/ilspy/ICSharpCode.Decompiler.Console/bin/Release/netcoreapp2.0# cd debian-x64/
root@161a8e6a2bd4:/shared/ilspy/ICSharpCode.Decompiler.Console/bin/Release/netcoreapp2.0/debian-x64# ls
ilspycmd	    ilspycmd.pdb		     libhostfxr.so
ilspycmd.deps.json  ilspycmd.runtimeconfig.dev.json  libhostpolicy.so
ilspycmd.dll	    ilspycmd.runtimeconfig.json      publish
root@161a8e6a2bd4:/shared/ilspy/ICSharpCode.Decompiler.Console/bin/Release/netcoreapp2.0/debian-x64# ./ilspycmd ilspycmd.dll

Unhandled Exception: System.ArgumentOutOfRangeException: Index was out of range. Must be non-negative and less than the size of the collection.
Parameter name: index
   at System.ThrowHelper.ThrowArgumentOutOfRange_IndexException()
   at ICSharpCode.Decompiler.UniversalAssemblyResolver.GetAssemblyInNetGac(AssemblyNameReference reference, ReaderParameters parameters) in C:\projects\ilspy\ICSharpCode.Decompiler\DotNetCore\UniversalAssemblyResolver.cs:line 309
   at ICSharpCode.Decompiler.UniversalAssemblyResolver.ResolveInternal(AssemblyNameReference name, ReaderParameters parameters) in C:\projects\ilspy\ICSharpCode.Decompiler\DotNetCore\UniversalAssemblyResolver.cs:line 133
   at ICSharpCode.Decompiler.TypeSystem.DecompilerTypeSystem..ctor(ModuleDefinition moduleDefinition) in C:\projects\ilspy\ICSharpCode.Decompiler\TypeSystem\DecompilerTypeSystem.cs:line 57
   at ICSharpCode.Decompiler.Console.Program.Decompile(String assemblyFileName, TextWriter output, String typeName) in /shared/ilspy/ICSharpCode.Decompiler.Console/Program.cs:line 96
   at ICSharpCode.Decompiler.Console.Program.<>c__DisplayClass0_0.<Main>b__0() in /shared/ilspy/ICSharpCode.Decompiler.Console/Program.cs:line 70
   at McMaster.Extensions.CommandLineUtils.CommandLineApplication.Execute(String[] args)
   at ICSharpCode.Decompiler.Console.Program.Main(String[] args) in /shared/ilspy/ICSharpCode.Decompiler.Console/Program.cs:line 73
Aborted
root@161a8e6a2bd4:/shared/ilspy/ICSharpCode.Decompiler.Console/bin/Release/netcoreapp2.0/debian-x64#

**Comment by chrisdoman** [comment]
Oh wait... it works now :) 

I think I must have built an older version without realising
